AZ3166/src/libraries/AzureIoT/examples/FirmwareOTA/FirmwareOTA.ino

```diff
@@ -0,0 +1,164 @@
+  Serial.begin(115200);
+
+  // Initialize the WiFi module
+  Screen.print(3, " > WiFi");
+  hasWifi = false;
+  InitWifi();
+  if (!hasWifi)
+  {
+    return;
+  }
+
+  Screen.print(3, " > IoT Hub");
+  DevKitMQTTClient_SetOption(OPTION_MINI_SOLUTION_NAME, "FirmwareOTA");
+  DevKitMQTTClient_Init(true);
+  IoTHubClient_SetCurrentFwInfo(currentFirmwareVersion);
+  IoTHubClient_ReportOTAStatus(currentFirmwareVersion, "current");
+}
+
+void loop()
+{
+  if (hasWifi && enabledOTA)
+  {
+    DevKitMQTTClient_Check();
+    FW_INFO* fwInfo = new FW_INFO;
+    bool hasNewOTA = IoTHubClient_OTAHasNewFw(fwInfo);
+    if (hasNewOTA) {
+      if (strlen(fwInfo -> fwPackageURI) >= 6 && (strncmp("https:", fwInfo -> fwPackageURI, 6) != 0)) {
+        IoTHubClient_ReportOTAStatus(currentFirmwareVersion, "error", fwInfo -> fwVersion, "URINotHTTPS");
+      } else {
+        if (IoTHubClient_OTAVersionCompare(fwInfo -> fwVersion, currentFirmwareVersion) == 1) {
+          Screen.clean();
+          Screen.print(0, "New firmware.");
+          Screen.print(1, fwInfo -> fwVersion);
+          LogInfo(fwInfo -> fwVersion);
+          LogInfo(fwInfo -> fwPackageURI);
+          LogInfo(fwInfo -> fwPackageCheckValue);
+          LogInfo("%d", fwInfo -> fwSize);
+
+          IoTHubClient_ReportOTAStatus(currentFirmwareVersion, "downloading", fwInfo -> fwVersion, NULL, getTimeStamp());
+          DevKitMQTTClient_Close();
+          Screen.clean();
+          Screen.print("Downloading...");
+          OTAUpdateClient& otaClient = OTAUpdateClient::getInstance();
+          int result = otaClient.updateFromUrl(fwInfo -> fwPackageURI, httpsCert);
+          if (result == 0) {
+            Screen.print("Download success\nVeryifying...\n");
+            DevKitMQTTClient_Init(true);
+            IoTHubClient_ReportOTAStatus(currentFirmwareVersion, "verifying", fwInfo -> fwVersion);
+            
+            static volatile uint32_t flashIdx = 0;
+            uint8_t *checkBuffer = (uint8_t *)malloc(1);
+            int i = fwInfo -> fwSize;
+            uint16_t checkSum = 0;
+            CRC16_Context contex;
+            CRC16_Init(&contex);
+            while (i--) {
+              MicoFlashRead((mico_partition_t)MICO_PARTITION_OTA_TEMP, &flashIdx, checkBuffer, 1);
+              CRC16_Update(&contex, checkBuffer, 1);
+            }
+            free(checkBuffer);
+            CRC16_Final(&contex, &checkSum);
+            char *checkSumString = checksumToString(checkSum);
+            LogInfo("CRC16 result: %d, %s", checkSum, checkSumString);
+            Screen.clean();
+            if (strcmp(checkSumString, fwInfo -> fwPackageCheckValue) == 0) {
+              Screen.print("Verify success\n");
+              LogInfo("Verify success");
+              IoTHubClient_ReportOTAStatus(currentFirmwareVersion, "applying", fwInfo -> fwVersion, NULL, NULL, getTimeStamp());
+              mico_system_reboot();
+            } else {
+              Screen.print("Verify failed\n");
```

Comment:
verify failed and download failed are very similar. Consider combine to a function.